docs: Update HA Terraform reference and add starter cluster reference (#32354)
gravitational/teleport: docs/pages/deploy-a-cluster/deployments/aws-starter-cluster-terraform.mdx

---
title: Teleport Single Instance deployment on AWS
description: How to quickly configure Teleport on a single instance for testing in AWS.
tocDepth: 3
---

> This guide is designed to accompany our [reference starter-cluster Terraform code](https://github.com/gravitational/teleport/tree/branch/v(=teleport.major_version=)/examples/aws/terraform/starter-cluster#teleport-terraform-aws-ami-simple-example)

Reviewer: I would briefly describe the infrastructure components that this Terraform module deploys so a user can determine whether the module and this guide are appropriate for their use case.

and describe how to manage the resulting Teleport deployment.

(!docs/pages/includes/cloud/call-to-action.mdx!)

## Prerequisites

Our code requires Terraform 0.13+. You can [download Terraform here](https://www.terraform.io/downloads.html). We will assume that you have
`terraform` installed and available on your path.

```code
$ which terraform
/usr/local/bin/terraform
$ terraform version
Terraform v1.5.6
```

You will also require the `aws` command line tool. This is available in Ubuntu/Debian/Fedora/CentOS and MacOS Homebrew
as the `awscli` package.

Fedora/CentOS: `yum -y install awscli`

Ubuntu/Debian: `apt-get -y install awscli`

MacOS (with [Homebrew](https://brew.sh/)): `brew install awscli`

When possible, installing via a package is always preferable. If you can't find a package available for
your distribution, you can also download the tool from [https://aws.amazon.com/cli/](https://aws.amazon.com/cli/)

> We will assume that you have configured your AWS cli access with credentials available at `~/.aws/credentials`:

Reviewer: ```suggestion
We will assume that you have configured your AWS CLI access with credentials available at `~/.aws/credentials`:
```

Author: We already state this 20 lines below when we talk about `awscli` - was this intentional?

commands will detect it and remind you to do so if necessary.
```

This will download the appropriate Terraform plugins needed to spin up Teleport using our
reference code.

## Set up variables

Terraform modules use variables to pass in input. You can do this in a few ways:

- on the command line to `terraform apply`
- by editing the `vars.tf` file
- by setting environment variables

For this guide, we are going to make extensive use of environment variables. This is because it makes it easier for us
to reference values from our configuration when running Teleport commands after the cluster has been created.

Any set environment variable starting with `TF_VAR_` is automatically processed and stripped down by Terraform, so
`TF_VAR_test_variable` becomes `test_variable`.

We maintain an up-to-date list of the variables and what they do in the README.md file under [the
`examples/aws/terraform/ha-autoscale-cluster` section of the Teleport repo](https://github.com/gravitational/teleport/blob/branch/v(=teleport.major_version=)/examples/aws/terraform/starter-cluster/README.md)
but we'll run through an example list here.

Things you will need to decide on:

### region

```code

> export TF_VAR_region="<Var name="us-west-2" />"

Reviewer: All of the `code` snippets on this page with example commands should start with `$` so the docs engine renders them properly

<Admonition
  type="tip"
  title="List Gravitational AMIs"
>
  OSS AMIs<br/>
  ```code
  aws --region <Var name="us-west-2" /> ec2 describe-images --owners 126027368216 --filters 'Name=name,Values=gravitational-teleport-ami-oss-(=teleport.version=)'
  ```

  Enterprise AMIs<br/>
  ```code
  aws --region <Var name="us-west-2" /> ec2 describe-images --owners 126027368216 --filters 'Name=name,Values=gravitational-teleport-ami-ent-(=teleport.version=)'
  ```

  List Enterprise FIPS 140-2 AMIs<br/>
  ```code
  aws --region <Var name="us-west-2" /> ec2 describe-images --owners 126027368216 --filters 'Name=name,Values=gravitational-teleport-ami-ent-(=teleport.version=)-fips'
  ```
</Admonition>

### key_name

```code
export TF_VAR_key_name="exampleuser"
```

The AWS keypair name to use when deploying EC2 instances. This must exist in the same region as you
specify in the `region` variable, and you will need a copy of this keypair available to connect to the deployed
EC2 instances. Do not use a keypair that you do not have access to.

### license_path

(!docs/pages/includes//enterprise/obtainlicense.mdx!)

```code
export TF_VAR_license_path="/home/user/license.pem"
```

> This license will be uploaded to AWS SSM and automatically downloaded to Teleport auth nodes in order to enable

Reviewer: ```suggestion
This license will be uploaded to AWS SSM and automatically downloaded to Teleport Auth Service nodes in order to enable
```

Author: This was inappropriate here anyway as it's a single-instance deployment - fixed.

> Teleport Enterprise/Pro functionality.

Reviewer: ```suggestion
Teleport Enterprise functionality.
```

> (OSS users can provide any valid local file path here - it isn't used by the auth server in a Teleport OSS install)

Reviewer: I would leave a single no-op file path that OSS users can use here so they don't need to come up with one (e.g., `./no-op` or whatever passes validation).

> (OSS users can provide any valid local file path here - it isn't used by the auth server in a Teleport OSS install)

Reviewer: ```suggestion
(OSS users can provide any valid local file path here - it isn't used by the Auth Service in a Teleport OSS install.)
```

with `-events` appended to the end.

In our example, the DynamoDB table would be called `teleport-example-events`.

More information about how Teleport works with DynamoDB can be found in our [Storage Backends guide](../../reference/backends.mdx#dynamodb).

### Recorded session storage

The reference Terraform deployment sets Teleport up to store recorded session logs in the same S3 bucket configured in
the [`s3_bucket_name`](#s3\_bucket_name) variable, under the `records` directory.

In our example this would be `s3://teleport-example/records`

<Admonition
  type="tip"
  title="Tip"
>
  S3 provides [Amazon S3 Object Lock](https://docs.aws.amazon.com/AmazonS3/latest/dev/object-lock.html),
  which is useful for customers deploying Teleport in regulated environments. Configuration of object lock is out of
  the scope of this guide.
</Admonition>

### Cluster domain

The reference Terraform deployment sets the Teleport cluster up to be available on a domain defined in Route 53, referenced
by the [`route53_domain`](#route53\_domain) variable. In our example this would be `teleport.example.com`.

> Teleport's web interface will be available on port 443 - [https://teleport.example.com](https://teleport.example.com).

Reviewer: I wouldn't link to the example URL.

```suggestion
Teleport's web interface will be available on port 443 of your Teleport cluster host.
```

When using ACM, this is an A record aliased to the AWS Application Load Balancer.

When using Let's Encrypt, this is an A record pointing to the instance's public IP.

> <Tabs>

Reviewer: Same comments on this `Tabs` box as the similar `Tabs` box in the other guide

# Terraform will perform the following actions:
#  <output trimmed>

Plan: 30 to add, 0 to change, 0 to destroy.

Changes to Outputs:
  + cluster_name        = "example-cluster"
  + cluster_web_address = "https://teleport.example.com"
  + instance_ip_public  = (known after apply)
  + key_name            = "example"

Do you want to perform these actions?
  Terraform will perform the actions described above.
  Only 'yes' will be accepted to approve.

  Enter a value: yes

------------------------------------------------------------------------

Note: You didn't specify an "-out" parameter to save this plan, so Terraform
can't guarantee that exactly these actions will be performed if
"terraform apply" is subsequently run.
```

This looks good (no errors produced by Terraform) so we can run `terraform apply`:

```code
$ terraform apply
#  <output trimmed>

# Plan: 30 to add, 0 to change, 0 to destroy.

# Do you want to perform these actions?
#  Terraform will perform the actions described above.
#  Only 'yes' will be accepted to approve.

#  Enter a value:
```

Entering `yes` here will start the Terraform deployment. It takes around 5 minutes to deploy in full.

> ### Destroy/shutdown a Terraform deployment

Reviewer: ```suggestion
### Destroy/shut down a Terraform deployment
```

If you need to tear down a running deployment for any reason, you can run `terraform destroy`.

## Accessing the cluster after Terraform setup

Once the Terraform setup is finished, the URL to your Teleport cluster's web UI will be set in the `cluster_web_address` Terraform output.

You can see this after deploy by running `terraform output -raw cluster_web_address`:

```code
$ terraform output -raw cluster_web_address
https://teleport.example.com
```

### Adding an admin user to the Teleport cluster

> To add users to the Teleport cluster, you will need to connect to a Teleport auth server via SSH and run the `tctl` command.

Reviewer: ```suggestion
To add users to the Teleport cluster, you will need to connect to a Teleport Auth Service via SSH and run the `tctl` command.
```

> 1 - Use the key name and server IP to SSH into the auth server. Make sure that the AWS keypair that you specified in the

Reviewer: We should use a standard Markdown ordered list here:

```
1. Item
1. Item
1. Item
```

> 1 - Use the key name and server IP to SSH into the auth server. Make sure that the AWS keypair that you specified in the

Reviewer: ```suggestion
1 - Use the key name and server IP to SSH into the Auth Service. Make sure that the AWS keypair that you specified in the
```
